frontends/PyCDE/src/pycde/testing.py

```diff
@@ -40,3 +44,132 @@ def testmodule_inner(func_or_class):
     return mod
 
   return testmodule_inner
+
+
+def pycdetest(func):
+  # Set a flag on the function to indicate that it's a testbench.
+  setattr(func, "_testbench", True)
+  return func
+
+
+def gen_cocotb_makefile(top, testmod, sources, sim):
+  """
+  Creates a simple cocotb makefile suitable for driving the testbench.
+  """
+  template = f"""
+TOPLEVEL_LANG = verilog
+VERILOG_SOURCES = {" ".join(list(sources))}
+TOPLEVEL = {top}
+MODULE = {testmod}
+SIM={sim}
+
+include $(shell cocotb-config --makefiles)/Makefile.sim
+"""
+  return template
+
+
+def gen_cocotb_testfile(tests):
+  """
+  Converts testbench functions to cocotb-compatible versions..
+  To do this cleanly, we need to detect the indent of the function,
+  and remove it from the function implementation.
+  """
+  template = "import cocotb\n\n"
+
+  for test in tests:
+    src = inspect.getsourcelines(test)[0]
+    indent = len(src[0]) - len(src[0].lstrip())
+    src = [line[indent:] for line in src]
+    # Remove the '@pycdetest' decorator
+    src = src[1:]
+    # If the function was not async, make it so.
+    if not src[0].startswith("async"):
+      src[0] = "async " + src[0]
+
+    # Append to the template as a cocotb test.
+    template += "@cocotb.test()\n"
+    template += "".join(src)
+    template += "\n\n"
+
+  return template
+
+
+class IVerilogHandler:
```

Comment:
Does cocotb not do this?

Reply:
Calling `iverilog` yes, versioning no. I can remove it but I think a user might want to have a warning system for `iverilog` versioning instead of getting fairly obscure errors from `cocotb` executing and old version of `iverilog`.